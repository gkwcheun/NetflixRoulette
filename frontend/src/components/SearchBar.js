import React from "react";
import { connect } from "react-redux";
import { fetchMovies } from "../redux/movie/movieActions";
import { useEffect, useState } from "react";
import { useLocation, useHistory } from "react-router-dom";
import PropTypes from "prop-types";

function SearchBar(props) {
  const { fetchMovies } = props;
  const [search, setSearch] = useState("");
  const [searchQuery, setSearchQuery] = useState({
    query: null,
    category: null,
    sortBy: null,
    sortOrder: null,
  });
  const history = useHistory();

  const submitSearch = (e) => {
    //   refactor later for api call
    e.preventDefault();
    history.push(`/search?title=${search}`);
    let searchInput = {
      query: search,
      category: "title",
      sortBy: null,
      sortOrder: null,
    };
    setSearchQuery(searchInput);
  };

  const query = new URLSearchParams(useLocation().search);
  console.log(query.get("sortBy"));

  let searchObj = {
    query: query.get("title") ? query.get("title") : query.get("genre"),
    category: query.get("title")
      ? "title"
      : query.get("genre")
      ? "genres"
      : null,
    sortBy: query.get("sortBy") ? query.get("sortBy") : null,
    sortOrder: query.get("sortOrder") ? query.get("sortOrder") : null,
  };

  useEffect(() => {
    setSearchQuery(searchObj);
  }, []);

  useEffect(() => {
    let { query, category, sortBy, sortOrder } = searchQuery;
    if (query) {
      fetchMovies(
        null,
        sortBy,
        sortOrder,
        query.toLowerCase(),
        category.toLowerCase()
      );
      setSearch(query);
    }
  }, [searchQuery]);

  return (
    <div className="search-container">
      <h1 className="find-movie-header">FIND YOUR MOVIE</h1>
      <div>
        <input
          className="search-input"
          type="text"
          placeholder="   What do you want to watch?"
          onChange={(e) => setSearch(e.target.value)}
          value={search}
        ></input>
        <button
          name="search-button"
          data-testid="movie-search-btn"
          className="search-btn"
          onClick={submitSearch}
        >
          Search
        </button>
      </div>
    </div>
  );
}

const mapDispatchToProps = (dispatch) => {
  return {
    fetchMovies: (
      genre = null,
      criteria = null,
      order = null,
      query = null,
      category = null
    ) => {
      return dispatch(fetchMovies(genre, criteria, order, query, category));
    },
  };
};

SearchBar.propTypes = {
  fetchMovies: PropTypes.func.isRequired,
};

export default connect(null, mapDispatchToProps)(SearchBar);
