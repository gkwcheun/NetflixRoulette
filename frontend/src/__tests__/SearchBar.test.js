import React from "react";
import configureStore from "redux-mock-store";
import renderer from "react-test-renderer";
import { Provider } from "react-redux";
import { render, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import SearchBar from "../components/SearchBar";

const mockStore = configureStore([]);
const mockHistoryPush = jest.fn();
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
  useLocation: () => ({
    search: "",
  }),
}));

test("placeholder", () => {});

describe("Search Bar Testing", () => {
  let store;
  let componentSnapshot;
  let component;
  let initialState = {};
  beforeEach(() => {
    store = mockStore(initialState);
    store.dispatch = jest.fn();
    componentSnapshot = renderer.create(
      <Provider store={store}>
        <SearchBar />
      </Provider>
    );
    component = render(
      <Provider store={store}>
        <SearchBar />
      </Provider>
    );
  });
  test("component renders with given state from redux store", () => {
    expect(componentSnapshot.toJSON()).toMatchSnapshot();
  });
  test("Verify search bar title was rendered", () => {
    const { getByText } = component;
    getByText("FIND YOUR MOVIE");
  });
  test("Verify Search Input Rendered", () => {
    const { getByPlaceholderText } = component;
    getByPlaceholderText("What do you want to watch?");
  });
  test("Changing input in search bar works correctly", () => {
    const { getByPlaceholderText } = component;
    const searchInput = getByPlaceholderText("What do you want to watch?");
    fireEvent.change(searchInput, {
      target: {
        value: "La La Land",
      },
    });
    expect(searchInput.value).toBe("La La Land");
  });
  test("Verify seach button renders", () => {
    const { getByTestId } = component;
    getByTestId("movie-search-btn");
  });
  test("Test each button click pushes to url", () => {
    const { getByTestId } = component;
    const searchBtn = getByTestId("movie-search-btn");
    fireEvent.click(searchBtn);
    expect(mockHistoryPush).toHaveBeenCalledTimes(1);
  });
  test("Change url input and push search to url", () => {
    const { getByPlaceholderText } = component;
    const searchInput = getByPlaceholderText("What do you want to watch?");
    fireEvent.change(searchInput, {
      target: {
        value: "La La Land",
      },
    });
    expect(searchInput.value).toBe("La La Land");
    const { getByTestId } = component;
    const searchBtn = getByTestId("movie-search-btn");
    fireEvent.click(searchBtn);
    expect(mockHistoryPush).toHaveBeenCalledWith(
      `/search?title=${"La La Land"}`
    );
  });
});
