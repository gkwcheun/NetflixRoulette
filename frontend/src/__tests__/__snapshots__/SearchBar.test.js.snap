// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Bar Testing component renders with given state from redux store 1`] = `
<div
  className="search-container"
>
  <h1
    className="find-movie-header"
  >
    FIND YOUR MOVIE
  </h1>
  <div>
    <input
      className="search-input"
      onChange={[Function]}
      placeholder="   What do you want to watch?"
      type="text"
      value=""
    />
    <button
      className="search-btn"
      data-testid="movie-search-btn"
      name="search-button"
      onClick={[Function]}
    >
      Search
    </button>
  </div>
</div>
`;
